clippy_lints/src/matches/single_match.rs

```diff
@@ -122,133 +124,231 @@ fn report_single_pattern(
+                            return matches!(self, Self::Wild);
+                        };
+                        let is_wild = if let [pat] = pats
+                            && variant.fields.len() == 1
+                        {
+                            state.add_pat(cx, pat)
+                        } else {
+                            state.add_and_pats(cx, pats)
+                        };
+                        is_wild && self.check_all_wild_enum()
+                    },
+                    _ => matches!(self, Self::Wild),
                 }
-                if found_dotdot {
-                    continue;
+            },
+            PatKind::Struct(ref path, pats, _) => {
+                let Some(adt) = cx.typeck.pat_ty(pat).ty_adt_def() else {
+                    return matches!(self, Self::Wild);
+                };
+                if adt.is_struct() {
+                    return if let [pat] = pats
+                        && adt.non_enum_variant().fields.len() == 1
+                    {
+                        self.add_pat(cx, pat.pat)
+                    } else {
+                        self.add_and_pats(cx, pats.iter().map(|x| x.pat))
+                    };
                 }
-                if !contains_only_wilds(&left_in[i - left_dot_space])
-                    && !contains_only_wilds(&right_in[i - right_dot_space])
-                {
-                    return false;
+                match cx.tcx.get_diagnostic_name(adt.did()) {
+                    Some(sym::Option) => {
+                        if let [pat] = pats {
+                            self.add_pat(cx, pat.pat)
+                        } else {
+                            *self = Self::Wild;
+                            true
+                        }
+                    },
+                    Some(sym::Result | sym::Cow) => {
+                        let Some((state, variant)) = self.get_enum_variant(cx, adt, path, pat.hir_id) else {
+                            return matches!(self, Self::Wild);
+                        };
+                        let is_wild = if let [pat] = pats
+                            && variant.fields.len() == 1
+                        {
+                            state.add_pat(cx, pat.pat)
+                        } else {
+                            state.add_and_pats(cx, pats.iter().map(|x| x.pat))
+                        };
+                        is_wild && self.check_all_wild_enum()
+                    },
+                    _ => matches!(self, Self::Wild),
                 }
-            }
-            true
-        },
-        (PatKind::TupleStruct(..), PatKind::Path(_)) => pat_in_candidate_enum(cx, ty, right),
-        (PatKind::TupleStruct(..), PatKind::TupleStruct(_, inner, _)) => {
-            pat_in_candidate_enum(cx, ty, right) && inner.iter().all(contains_only_wilds)
-        },
-        _ => false,
+            },
+        }
     }
```

Comment:
Could you add some comments here? I'm having trouble understanding the meaning behind all these branches (and the meaning behind the function in general)